node/pkg/driver/device_connectivity/device_connectivity_helper_scsigeneric.go

```diff
@@ -128,117 +130,52 @@ func (r OsDeviceConnectivityHelperScsiGeneric) RescanDevices(lunId int, arrayIde
 
-	if len(devicePaths) == 0 && len(errStrings) != 0 {
-		err := errors.New(strings.Join(errStrings, GetMpahDevErrorsSep))
+	if err != nil {
 		return "", err
 	}
-
-	devicePathTosysfs := make(map[string]bool)
-	// Looping over the physical devices of the volume - /dev/sdX and store all the dm devices inside map.
-	for _, path := range devicePaths {
-		if path != "" { // since it may return empty items
-			mappedDevicePath, err := r.Helper.GetMultipathDisk(path)
-			if err != nil {
-				return "", err
-			}
-
-			if mappedDevicePath != "" {
-				devicePathTosysfs[mappedDevicePath] = true // map it in order to save uniq dm devices
-			}
-
-		}
+	if mpathdOutput == "" {
+		return "", &MultipathDeviceNotFoundForVolumeError{volumeId}
 	}
-
-	var mps string
-	for key := range devicePathTosysfs {
-		mps += ", " + key
+	dms := make(map[string]bool)
+	scanner := bufio.NewScanner(strings.NewReader(mpathdOutput))
+	for scanner.Scan() {
+		deviceLine := scanner.Text()
+		lineParts := strings.Split(deviceLine, MpathdSeparator)
+		dm, uuid := lineParts[0], lineParts[1]
+		if strings.Contains(uuid, volumeUuidLower) {
+			dmPath := filepath.Join(DevPath, filepath.Base(dm))
+			dms[dmPath] = true
+			logger.Infof("GetMpathDevice: DM found: %s for volume %s", dmPath, uuid)
+		}
 	}
-	logger.Infof("GetMpathDevice: Found multipath devices: [%s] that relats to lunId=%d and arrayIdentifiers=%s", mps, lunId, arrayIdentifiers)
 
-	if len(devicePathTosysfs) > 1 {
-		return "", &MultipleDmDevicesError{volumeId, lunId, arrayIdentifiers, devicePathTosysfs}
+	if len(dms) > 1 {
+		return "", &MultipleDmDevicesError{volumeId, dms}
 	}
 
 	var md string
-	for md = range devicePathTosysfs {
+	for md = range dms {
 		break // because its a single value in the map(1 mpath device, if not it should fail above), so just take the first
 	}
-	return md, nil
-}
 
-func (r OsDeviceConnectivityHelperScsiGeneric) waitForMpath(targetPath string, connectivityType string, arrayIdentifier string, lunId int, volumeId string,
-	resChannel chan<- *WaitForMpathResult) {
-	lunIdStr := convertIntToScsilun(lunId)
-	dp := targetPath + strings.Join([]string{connectivityType, arrayIdentifier, "lun", lunIdStr}, "-")
-	logger.Infof("waitForMpath: Get the mpath devices related to connectivityType=%s initiator=%s and lunID=%d : {%v}", connectivityType, arrayIdentifier, lunId, dp)
-	dps, exists, e := r.Helper.WaitForPathToExist(dp, WaitForMpathRetries, WaitForMpathWaitIntervalSec)
-	if e != nil {
-		logger.Errorf("waitForMpath: No device found error : %v ", e.Error())
-	} else if !exists {
-		e = &MultipathDeviceNotFoundForLunError{volumeId, lunId, []string{arrayIdentifier}}
-		logger.Errorf(e.Error())
+	SqInqWwn, err := r.Helper.GetWwnByScsiInq(md)
+	if err != nil {
+		return "", err
+	}
+
+	if strings.ToLower(SqInqWwn) != volumeUuidLower {
+		// To make sure we found the right WWN, if not raise error instead of using wrong mpath
```

Comment:
shouldn't we scan the other DMs for the correct wwn?

Reply:
no, it should be a mach. even if we found our volume uuid in another dm how can we be sure its a good mount?

Reply:
by running `sg_inq` with the other dm?

Reply:
but if there is no match can we say fore sure that sg_inq is correct? (and on what dm are we gonna do `sg_inq`?)

Reply:
no, why not just `sg_inq` every dm that shows our uuid until there's a match?

Reply:
there wont be 2 dms with same uuid.. AFAIK

Reply:
and if no so I cant see what the scanning will give.

Reply:
as we saw [here ](https://github.com/IBM/ibm-block-csi-driver/blob/eb4fe0d6afb46a26d74835c458302c839e9e6312/node/pkg/driver/device_connectivity/device_connectivity_helper_scsigeneric.go#L476), it is not relevant, we wont have 2 dms with the same uuid on this part..